Sources/TensorFlow/Optimizers/MomentumBased.swift

```diff
@@ -162,17 +162,101 @@ public class AdaDelta<Model: Differentiable>: Optimizer
+///   - beta2: A Float. The exponentian decay rate for the 2nd moment estimates (default value: 
+///     0.999).
+///   - epsilon: A Float. A small scalar added to the denominator to improve numerical stability 
+///     (default value: 1e-8).
+///   - decay: A Float. The learning rate decay (default value: 0).
+///
+/// ### Examples: ###
+///
+/// - Train a simple reinforcement learning agent:
+///
+/// ````
+/// ...
+/// // Instantiate an agent's policy - approximated by the neural network (`net`) after defining it 
+/// in advance.
+/// var net = Net(observationSize: Int(observationSize), hiddenSize: hiddenSize, actionCount: actionCount)
+/// // Define the Adam optimizer for the network with a learning rate set to 0.01.
+/// let optimizer = Adam(for: net, learningRate: 0.01)
+/// ...
+/// // Begin training the agent (over a certain number of episodes).
+/// while true {
+/// ...
+///     // Implementing the gradient descent with the Adam optimizer:
+///     // Define the gradients (use withLearningPhase to call a closure under a learning phase).
+///     let gradients = withLearningPhase(.training) {
+///         TensorFlow.gradient(at: net) { net -> Tensor<Float> in
+///             // Return a softmax (loss) function
+///             return loss = softmaxCrossEntropy(logits: net(input), probabilities: target)
+///         }
+///     }
+///     // Update the differentiable variables of the network (`net`) along the gradients with the Adam 
+/// optimizer.
+///     optimizer.update(&net, along: gradients)
+///     ...
+///     }
+/// }
+/// ````
+///
+/// - Train a generative adversarial network (GAN):
+///
+/// ````
+/// ...
+/// // Instantiate the generator and the discriminator networks after defining them.
+/// var generator = Generator()
+/// var discriminator = Discriminator()
+/// // Define the Adam optimizers for each network with a learning rate set to 2e-4 and beta1 - to 0.5.
+/// let adamOptimizerG = Adam(for: generator, learningRate: 2e-4, beta1: 0.5)
+/// let adamOptimizerD = Adam(for: discriminator, learningRate: 2e-4, beta1: 0.5)
+/// ...
+/// Start the training loop over a certain number of epochs (`epochCount`).
+/// for epoch in 1...epochCount {
+///     // Start the training phase.
+///     ...
+///     for batch in trainingShuffled.batched(batchSize) {
+///         // Implementing the gradient descent with the Adam optimizer:
+///         // 1) Update the generator.
+///         ...
+///         let 𝛁generator = TensorFlow.gradient(at: generator) { generator -> Tensor<Float> in
+///             ...
+///             return loss
+///             }
+///         // Update the differentiable variables of the generator along the gradients (`𝛁generator`) 
+///         // with the Adam optimizer.
+///         adamOptimizerG.update(&generator, along: 𝛁generator)
+///
+///         // 2) Update the discriminator.
+///         ...
+///         let discriminator = TensorFlow.gradient(at: generator) { generator -> Tensor<Float> in
```

Comment:
let **𝛁discriminator** = TensorFlow.gradient(at: **discriminator**) { **discriminaotr** -> Tensor<Float> in